### ISSUE TITLE
Inception Score calculation

### ISSUE THREAD

**Issue by sbarratt** [openai/improved-gan#29]
The Inception Score calculation has 3 mistakes.

### It uses an outdated Inception network that in fact outputs a 1008-vector of classes (see [the following GitHub issue](https://github.com/tensorflow/tensorflow/issues/4128)):

> It turns out that the 1008 size softmax output is an artifact of dimension back-compatibility with a older, Google-internal system. Newer versions of the inception model have 1001 output classes, where one is an "other" class used in training. You shouldn't need to pay any attention to the extra 8 outputs.

**Fix**: See [link](https://research.googleblog.com/2016/08/improving-inception-and-image.html) for the new inception Model.

### It calculates the kl-divergence directly using logs, which leads to numerical instabilities (can output nan instead of inf). Instead, `scipy.stats.entropy` should be used.

```python
kl = part * (np.log(part) - np.log(np.expand_dims(np.mean(part, 0), 0)))
kl = np.mean(np.sum(kl, 1))
``` 

**Fix**: Replace the above with something along the lines of the following:
```python
py = np.mean(part, axis=0)
l = np.mean([entropy(part[i, :], py) for i in range(part.shape[0])])
```

### It calculates the mean of the exponential of the split rather than the exponential of the mean:

Here is the code in `inception_score.py` which does this:
```python
    scores.append(np.exp(kl))
return np.mean(scores), np.std(scores)
```

This is clearly problematic, as can easily be seen in a very simple case with a x~Bernoulli(0.5) random variable that E[e^x] = .5(e^(0) + e^(1)) != e^(.5(0)+.5(1)) = e^[E[x]]. This can further be seen with an example w/ a uniform random variable, where the split-mean over-estimates the exponential.

```python
import numpy as np
data = np.random.uniform(low=0., high=15., size=1000)
split_data = np.split(data, 10)
np.mean([np.exp(np.mean(x)) for x in split_data]) # 1608.25
np.exp(np.mean(data)) # 1477.25
```

**Fix**: Do not calculate the mean of the exponential of the split, and instead calculate the exponential of the mean of the KL-divergence over all 50,000 inputs.

**Comment by xunhuang1995** [comment]
The first point is an important issue. For the third point, note that they do NOT intend to calculate the inception score (IS) for 50,000 inputs. Rather, they spit 50,000 samples into 10 splits each with 5,000 samples. They then calculate IS for each split and return the average IS over splits. So the code is correct.

**Comment by mathfinder** [comment]
So,which version is correct？

**Comment by lipanpeng** [comment]
Can anyone tell me where can I find the material about the inception score?

**Comment by christopher-beckham** [comment]
Not entirely relevant to the discussion but a recent commit to that file made a minor fix by changing the batch size from 100 to 1, which seems to me inefficient considering you may need to compute it for 50k samples. I don't use TensorFlow (though I need to run this code) so I'm just curious as to why a batch size >1 throws an error (the error seems to imply the graph can only take inputs of size `(1, ?, ?, 3)`).
